src/Paramore.Brighter.Outbox.MsSql/MsSqlOutbox.cs

```diff
@@ -555,177 +295,207 @@ public void MarkDispatched(Guid id, DateTime? dispatchedAt = null, Dictionary<st
-        public async Task<IEnumerable<Message>> OutstandingMessagesAsync(
+        public Task<IEnumerable<Message>> OutstandingMessagesAsync(
            double millSecondsSinceSent, 
            int pageSize = 100, 
            int pageNumber = 1,
            Dictionary<string, object> args = null,
            CancellationToken cancellationToken = default)
        {
-           var connection = await _connectionProvider.GetConnectionAsync(cancellationToken);
-           using (var command = connection.CreateCommand())
-           {
-               CreatePagedOutstandingCommand(command, millSecondsSinceSent, pageSize, pageNumber);
-
-               if(connection.State!= ConnectionState.Open) await connection.OpenAsync(cancellationToken);
-
-               if (_connectionProvider.HasOpenTransaction) command.Transaction = _connectionProvider.GetTransaction(); 
-               var dbDataReader = await command.ExecuteReaderAsync(cancellationToken);
-
-               var messages = new List<Message>();
-               while (await dbDataReader.ReadAsync(cancellationToken))
-               {
-                   messages.Add(MapAMessage(dbDataReader));
-               }
-               dbDataReader.Close();
-                
-               if(!_connectionProvider.IsSharedConnection) connection.Dispose();
-               else if (!_connectionProvider.HasOpenTransaction) connection.Close();
-                
-               return messages;
-           }
+            return RunCommandAsync(connection => CreatePagedOutstandingCommand(connection, millSecondsSinceSent, pageSize, pageNumber), dr => MapListFunctionAsync(dr, cancellationToken), cancellationToken);
        }
-        
-        private void CreatePagedDispatchedCommand(SqlCommand command, double millisecondsDispatchedSince, int pageSize, int pageNumber)
-        {
-            var pagingSqlFormat = "SELECT * FROM (SELECT ROW_NUMBER() OVER(ORDER BY Timestamp DESC) AS NUMBER, * FROM {0}) AS TBL WHERE DISPATCHED IS NOT NULL AND DISPATCHED < DATEADD(millisecond, @OutStandingSince, getutcdate()) AND NUMBER BETWEEN ((@PageNumber-1)*@PageSize+1) AND (@PageNumber*@PageSize) ORDER BY Timestamp DESC";
-            var parameters = new[]
-            {
-                CreateSqlParameter("PageNumber", pageNumber),
-                CreateSqlParameter("PageSize", pageSize),
-                CreateSqlParameter("OutstandingSince", -1 * millisecondsDispatchedSince)
-            };
-
-            var sql = string.Format(pagingSqlFormat, _configuration.OutBoxTableName);
 
-            command.CommandText = sql;
-            command.Parameters.AddRange(parameters);
-        }
+        #endregion
 
-        private void CreatePagedReadCommand(SqlCommand command, int pageSize, int pageNumber)
+        #region Things that Actually call SQL
+        private void RunNonQuery(IAmABoxTransactionConnectionProvider transactionConnectionProvider, Func<SqlConnection, SqlCommand> commandFunc, Action loggingAction)
```

Comment:
Isn't NonQuery a Command. This seems like a difficult negative naming choice

Reply:
Good spot, I can get that renamed

Reply:
I've renamed them ReadFromStore and WriteToStore Thought @ICooper ?